feat(store): enable dispatching actions on signal changes (#4600)
ngrx/platform: projects/ngrx.io/content/guide/store/actions.md

</div>

> ## Dispatching functions returning actions using signals

Author: Yes, that's much better. I've struggled finding the right title.

> There is also the option to dispatch a Signal of type `Signal<Action>`:

Reviewer: There are no Signal Actions anymore. This docs section needs to be updated to match the new implementation.

<code-example header="book.component.ts">
class BookComponent {
  bookId = input.required&lt;number&gt;();

  constructor(store: Store) {
    store.dispatch(() => loadBook({ id: this.bookId() })));
  }
}
</code-example>

`dispatch` executes initially and every time the `bookId` changes. If `dispatch` is called within an injection context, the signal is tracked until the context is destroyed. In the example above, that would be when `BookingComponent` is destroyed.

If `dispatch` is called outside an injection context, the signal is tracked during the whole application lifecycle.

Alternatively, you can provide your own injection context:

<code-example header="book.component.ts">
class BookComponent {  
  bookId = input.required&lt;number&gt;();
  injector = inject(Injector);

> 
>   ngOnInit(store: Store) {

Reviewer: ```suggestion
  store = inject(Store);

  ngOnInit() {
```

>     store.dispatch(() => loadBook({ id: this.bookId() }), { injector: this.injector });

Reviewer 2: Do we also want to update the first example using `inject`?
```suggestion
    this.store.dispatch(() => loadBook({ id: this.bookId() }), { injector: this.injector });
```

    store.dispatch(() => loadBook({ id: this.bookId() }), {injector: this.injector});
  }
}
</code-example>

> `dispatch` returns an `EffectRef` if you provide a signal. You can manually destroy the effect by calling `destroy` on the `EffectRef`.

Reviewer: ```suggestion
When passing a function to the `dispatch` method, it returns an `EffectRef`. For manual cleanup, call the `destroy` method on the `EffectRef`:
```

<code-example header="book.component.ts">
class BookComponent {
  bookId = input.required&lt;number&gt;();
  loadBookEffectRef: EffectRef | undefined;

> 
>   ngOnInit(store: Store) {

Reviewer: ```suggestion
  store = inject(Store);

  ngOnInit() {
```

    // uses the injection context of Store, i.e. root injector

>     this.loadBookEffectRef = store.dispatch(() => loadBook({ id: this.bookId() }));

Reviewer 2: ```suggestion
    this.loadBookEffectRef = this.store.dispatch(() => loadBook({ id: this.bookId() }));
```

</div>

## Signal Actions

There is also the option to dispatch a Signal of type `Signal<Action>`:

<code-example header="book.component.ts">

> class BookComponent {
>   bookId = input.required&lt;number&gt;();
>   loadBookAction = computed(() => loadBook({ id: this.bookId() }));
> 
>   constructor(store: Store) {
>     store.dispatch(this.loadBookAction);
>   }
> }

Reviewer: Instead of creating a computed signal, wouldn't it be simpler to provide a callback (that will be used as `effectFn` under the hood) to the `dispatch` method:

```ts
```suggestion
class BookComponent {
  bookId = input.required&lt;number&gt;();

  constructor(store: Store) {
    this.store.dispatch(() => loadBook({ id: this.bookId() }));
  }
}
```

Author: I agree, but then I'd remove. the option to pass a Signal. I don't see any use in creating a Signal containing an action.

Author: >  I'd remove the option to pass a Signal

Never mind 😅. Signal is also just a function that returns an Action.

Author: Done, see comments below

</code-example>

> `dispatch` executes initially and every time the `bookId` changes. If `dispatch` is called within an injection context, the Signal will be tracked until context is destroyed. In the example above, that would be when `BookingComponent` is destroyed.

Reviewer 2: ```suggestion
`dispatch` executes initially and every time the `bookId` changes. If `dispatch` is called within an injection context, the Signal is tracked until the context is destroyed. In the example above, that would be when `BookingComponent` is destroyed.
```

> If `dispatch` is called outside an injection context, the Signal will be tracked during the whole application lifecycle.

Reviewer 2: ```suggestion
If `dispatch` is called outside an injection context, the Signal is tracked during the whole application lifecycle.
```

Alternatively, you can provide your own injection context:

<code-example header="book.component.ts">
class BookComponent {  
  bookId = input.required&lt;number&gt;();
  loadBookAction = computed(() => loadBook({ id: this.bookId() }));
  injector = inject(Injector);

  ngOnInit(store: Store) {
    // runs outside the injection context
    store.dispatch(this.loadBookAction, {injector: this.injector});
  }
}
</code-example>

> `dispatch` will return an `EffectRef` if you provide a Signal. You can manually destroy the effect by calling `destroy` on the `EffectRef`.

Reviewer 2: ```suggestion
`dispatch` returns an `EffectRef` if you provide a Signal. You can manually destroy the effect by calling `destroy` on the `EffectRef`.
```

<code-example header="book.component.ts">
class BookComponent {
  bookId = input.required&lt;number&gt;();
  loadBookAction = computed(() => loadBook({ id: this.bookId() }));
  loadBookEffectRef: EffectRef | undefined;

  ngOnInit(store: Store) {

>     // uses injection context of Store, i.e. root injector

Reviewer 2: ```suggestion
    // uses the injection context of Store, i.e. root injector
```
